packages/runtime/test-runtime-utils/src/mocksForRebasing.ts

```diff
@@ -0,0 +1,133 @@
+
+		// We've processed something, therefore the current batch has ended
+		this.clientSequenceNumber++;
+	}
+
+	public submit(messageContent: any, localOpMetadata: unknown) {
+		this.outbox.push({
+			content: messageContent,
+			localOpMetadata,
+		});
+
+		// Messages in the same batch will have the same clientSequenceNumber
+		return this.clientSequenceNumber;
+	}
+
+	public flush() {
+		while (this.outbox.length > 0) {
+			// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+			this.submitInternal(this.outbox.shift()!);
+		}
+	}
+
+	private submitInternal(message: InternalMessage) {
+		this.factory.pushMessage({
+			clientId: this.clientId,
+			clientSequenceNumber: this.clientSequenceNumber,
+			contents: message.content,
+			referenceSequenceNumber: this.referenceSequenceNumber,
+			type: MessageType.Operation,
+		});
+		this.addPendingMessage(message.content, message.localOpMetadata, this.clientSequenceNumber);
+	}
+
+	public rebase() {
+		const messagesToRebase = this.outbox.slice();
+		this.outbox.length = 0;
+
+		messagesToRebase.forEach((message) =>
+			this.dataStoreRuntime.reSubmit(message.content, message.localOpMetadata),
+		);
+	}
+}
+
+interface InternalMessage {
+	content: any;
+	localOpMetadata: unknown;
+}
+
+/**
+ * Specialized implementation of MockContainerRuntimeFactory for testing op rebasing.
+ */
+export class MockContainerRuntimeFactoryForRebasing extends MockContainerRuntimeFactory {
+	public createContainerRuntime(
+		dataStoreRuntime: MockFluidDataStoreRuntime,
+		overrides?: { minimumSequenceNumber?: number },
+	): MockContainerRuntimeForRebasing {
+		const containerRuntime = new MockContainerRuntimeForRebasing(
+			dataStoreRuntime,
+			this,
+			overrides,
+		);
+		this.runtimes.push(containerRuntime);
+		return containerRuntime;
+	}
+
+	private processMessage() {
+		if (this.messages.length === 0) {
+			throw new Error("Tried to process a message that did not exist");
+		}
+
+		// Explicitly JSON clone the value to match the behavior of going thru the wire.
+		const message = JSON.parse(
+			JSON.stringify(this.messages.shift()),
+		) as ISequencedDocumentMessage;
+
+		this.minSeq.set(message.clientId, message.referenceSequenceNumber);
+		// Messages from the same batch have the same sequence number
+		message.sequenceNumber = this.sequenceNumber;
+		message.minimumSequenceNumber = this.getMinSeq();
+		for (const runtime of this.runtimes) {
+			runtime.process(message);
+		}
+	}
+
+	public processOneMessage() {
+		// Increase the sequence number between batches
+		this.sequenceNumber++;
+		this.processMessage();
+	}
+
+	public processSomeMessages(count: number) {
+		// Increase the sequence number between batches
+		this.sequenceNumber++;
+		for (let i = 0; i < count; i++) {
+			this.processMessage();
+		}
+	}
+
+	public processAllMessages() {
+		// Increase the sequence number between batches
+		this.sequenceNumber++;
+		this.processSomeMessages(this.messages.length);
```

Comment:
This is wrong, we should evacuate the whole queue at this point (even if the queue grows while it's being evacuated).